sysconfig/cloudinit.go

```diff
@@ -384,7 +420,35 @@ func configureCloudInit(model *asserts.Model, opts *Options) (err error) {
 		// it and ensure that the ubuntu-seed config matches the config from the
 		// gadget if that exists
 		// for now though, just return
-		return nil
+		installOpts.Filter = true
+
+		// handle the various permutations for the datasources mentioned in the
+		// gadget
+		switch {
+		case gadgetDatasourcesRes.ExplicitlyNoneAllowed:
+			// no datasources were allowed, so set it to the empty list
+			installOpts.AllowedDatasources = nil
+		case gadgetDatasourcesRes.ExplicitlyAllowed != nil:
+			// only allow what the gadget explicitly allowed
+			installOpts.AllowedDatasources = gadgetDatasourcesRes.ExplicitlyAllowed
```

Comment:
I looked at our original notes and I'm a bit confused, I'm happy to discuss this again but I think we don't want to copy things from seed that are not MAAS, so here we want the intersection (which can be empty) of ExplicitlyAllowed and supportedFilteredDatasources, until we have a use case for something that is not MAAS for this?. Except to make a special case around NoCloud?

Reply:
Hmm I suppose I misunderstood your suggestion when we discussed as I thought it was that what the gadget says always trumps everything else, so if the gadget says `datasource_list: [foo, bar]` then we allow for config from ubuntu-seed for foo and bar (because again this is just for `grade: signed`.

Reply:
but the filtering we do is based on what make sense for MAAS atm, we didn't analyze other things, also for those other datasources they could in theory communicate securely with the source of the data while ubuntu-seed is never secure in that sense

Reply:
Well actually sorry I'm confused now again re-reading your response before pushing the changes. If the gadget mentions a datasource other than MAAS, what should we do with cloud-init config files on ubuntu-seed _that are not specific to a datasource_ like network config. Should we copy it or not? I originally had implemented it such that we still will copy generic config over even if the datasource allowed list was empty, but I can't tell from the doc or your response here if that's what you want to do or not. 

Essentially we have the following cases:
* Gadget mentions MAAS -> great, copy all the config over, filter it for MAAS
* Gadget doesn't mention MAAS -> great, copy all the config over, filter it for MAAS
* Gadget mentions something other than MAAS -> ???
* Gadget explicitly mentions not to allow any datasource -> ???

For the last two cases we have the following options:
1. Install no config files from ubuntu-seed at all, even if they are generic and not specific to a datasource
2. Install _only_ config files that are generic and not specific to a datasource

Or we could do option 1 for the last case where the gadget says it wants no datasources, and option 2 for the case where the gadget says some other datasource.

Config that is not specific to i.e. MAAS can look like this:

```yaml
#cloud-config
network:
  config:
  - id: enp3s0
    mac_address: 52:54:00:b4:9e:25
    mtu: 1500
    name: enp3s0
    subnets:
    - address: 172.16.99.7/24
      dns_nameservers:
      - 172.16.99.1
      type: static
    type: physical
  - address: 172.16.99.1
    type: nameserver
  version: 1
```